### PR TITLE
ci: Out of the tarpit

### PR DESCRIPTION
**Pull request by MarcoPolo**
This PR gets us out of the tarpit of flaky tests. It reruns failed tests to see if they are flaky. If they consistently fail, the job fails. If they are flaky, they are recorded as such, but the job succeeds. The job outputs a sqlite database of all test runs. The table includes information on whether the test passed or failed and how many times for each.

The summary now includes what specific test failed, so you don't have to hunt through the logs to see what failed. It will also link you to an observable dashboard that lets you dive deeper. [example](https://github.com/libp2p/go-libp2p/actions/runs/10410612766)

I think it's worth experimenting with this, and we can always revert it later.

### Some valid concerns:

Q: Why do this?
A: Making changes has been feeling like walking through a tarpit with regard to CI. And a bad CI desensitizes you to actual failures. CI failures should be very high signal and very low noise.

Q: Won't this hide flaky tests?
A: No more than we are already doing by rerunning the tests. I'd even argue it hides it less because all the flakiness is presented on a single page rather than spread out across runs.

Q: What happens when a test is so flaky it fails everytime?
A: Sounds like an easy test to fix. Most, if not all, of our flakes happen less than 20%.

Q: Does this replace the [Flaky Tests dashboard](https://marcopolo.github.io/FlakyTests/)?
A: Not yet. Although I have a FlakyTestsV2 dashboard that makes better use of the sqlite format. This method is much easier to iterate on and much faster than the original Flaky Tests Dashboard.
 
 Q: Why are you "forking" go-test-template.yml?
 A: There's no other way of doing the changes I want to do. I think go-libp2p can graduate to owning its own `go-test` workflow. I prefer the simple workflow calling composite actions strategy to the shared reusable workflow across many projects. For now, the changes are minimal between the in-tree go-test-template.yml and the upstream one. In the future I may remove even more things from it.
 
This PR isn't intended to be the final state. I'd like to clean up the summaries and dashboard a lot more. But I do think this current branch is a lot better than what we have.
 
---

### How it works:

It introduces a `go test` helper script that:

1. Runs all the tests i.e. `go test ./...`
2. Saves the test results in a sqlite file.
3. If there are any failures it will rerun the failed tests `n` more times.
4. Uploads the sqlite db as an artifact of the job.

There's also an Observable dashboard that:

1. Parses all the sqlite db artifacts.
2. Shows which tests failed along with the relevant logs.
3. Derives a "flakiness score" which represents "for the given state of flakiness, how likely is a job to fail because of these flakes?". (n.b. It's a lot like rolling a dice with disadvantage.)

### CODE REVIEW

File: scripts/test_analysis/main.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	rows, err := db.QueryContext(ctx, `SELECT
+    tr_output.Package,
+    tr_output.Test,
+    GROUP_CONCAT(tr_output.Output,  "") AS Outputs
+FROM
+    test_results tr_fail
+JOIN
+    test_results tr_output
+ON
+    tr_fail.Test = tr_output.Test
```
Comment (sukunrt) [inline scripts/test_analysis/main.go:195]: Why do we need a self join here? Can't we just do a GROUP BY?

**Review by sukunrt — APPROVED** [review]
Can't wait to start using this :heart:

File: scripts/test_analysis/main.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	log.Printf("Found %d failed tests. Retrying them %d times", len(failedTests), retryCount)
+	hasOneNonFlakyFailure := false
+	loggedFlaky := map[string]struct{}{}
+
+	for _, ft := range failedTests {
+		isFlaky := false
+		for i := 0; i < retryCount; i++ {
+			log.Printf("Retrying %s.%s", ft.Package, ft.Test)
+			if err := t.goTestPkgTest(ft.Package, ft.Test, filterOutFlags(passThruFlags, coverRegex)); err != nil {
+				log.Printf("Failed to run %s.%s: %v", ft.Package, ft.Test, err)
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply scripts/test_analysis/main.go:68]: yes

File: scripts/test_analysis/cmd/gotest2sql/main.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Prepare the insert statement once
+	insertTime := time.Now().Format(time.RFC3339Nano)
+	stmt, err := tx.Prepare(`
+    INSERT INTO test_results (Time, Action, Package, Test, Elapsed, Output, BatchInsertTime)
+    VALUES (?, ?, ?, ?, ?, ?, ?)`)
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply scripts/test_analysis/cmd/gotest2sql/main.go:64]: I don't think so. Within a `go test` run `Package + Test` are unique. There are multiple rows for each `Package + Test`. Here's an example: https://observablehq.com/d/b3ba382153aea326

File: scripts/test_analysis/main.go (hunk 1 above)
Comment (MarcoPolo) [reply scripts/test_analysis/main.go:195]: same reason as https://github.com/libp2p/go-libp2p/pull/2923#discussion_r1723675521. Each row is just a different state/output line from `go test`. A single row doesn't represent the whole state of a single test.

**Pull request merged**